Added timeout/ping mechanism

hinrichsenhans opened the pull request:
If we haven't received a packet in 2500 ms, send a quick ping
If Eos doesn't send anything after a cumulative 5000ms, disconnect and reset the splash screen
This should address the cases where we've exited ETCnomad but the screen still shows the Pan/Tilt display.

This turns out to be a fair number of pings received as the console sits idle.

If we opened the filter up to receive another common OSC transmitted message from regular console use, we could avoid sending the ping so much. 

We could also change the ping interval to be closer to the timeout limit (4500 ms?), but ultimately it's the same problem if idle. Is this a concern?

hinrichsenhans committed:
Added timeout/ping mechanism

If we haven't received a packet in 2500 ms, send a quick ping
If Eos doesn't send anything after a cumulative 5000ms, disconnect
and reset the splash screen

samkearney commented:
I think this is a good feature. An OSC message every 2.5 seconds is hardly a bandwidth issue. @hinrichsenhans are you concerned about filling up the Eos logs with these pings over a long period of time?

samkearney commented on box_1/box_1_src/box_1_src.ino:
```diff
@@ -140,7 +140,8 @@ LiquidCrystal lcd(12, 11, 5, 4, 3, 2);
 bool updateDisplay = false;
 bool connectedToEos = false;
-
+uint32_t lastMessageRxTime = 0;
```
Nitpick: change to `unsigned long` to match return from `millis()`

samkearney commented on box_1/box_1_src/box_1_src.ino:
```diff
@@ -541,9 +543,37 @@ void loop()
+    {
+        uint32_t diff = millis() - lastMessageRxTime;
+        //We first check if it's been too long and we need to time out
+        if(diff > 5000) {
```
Nitpick: open brace on next line for style consistency

samkearney replied:
Consider changing 5000 and 2500 magic numbers to #defines

samkearney commented on box_1/box_1_src/box_1_src.ino:
```diff
@@ -541,9 +543,37 @@ void loop()
 
+    if(lastMessageRxTime > 0) 
+    {
+        uint32_t diff = millis() - lastMessageRxTime;
```
Nitpick: change to `unsigned long` to match return from `millis()`

samkearney reviewed (commented):
Looks good!

hinrichsenhans committed:
Updates from review

- Move magic numbers for timeouts to #defines
- Fix type mismatch
- Fix formatting

hinrichsenhans commented:
@bootress Yeah, 2.5 seconds is just a logging concern. We don't have OSC logging turned on by default in Eos, but if we turn it on for diagnostics, it's going to really start filling it up. (>17000 messages in a 12 hour period, like a console sitting idle overnight)
On the Arduino side, I suppose we could do something where we have an incremental back-off time, but the timeout is really trying to catch the random case of the console disappearing, not necessarily right after you stop touching it. 
On the Eos side, a heartbeat or adding an OSC command for us to send that isn't logged would be other solutions. 
All of these have trade-offs - ultimately I think this is a good starting point.

hinrichsenhans committed:
Version 1.0.0.5

Updated version string

samkearney commented:
@hinrichsenhans @MizPlusPlus My preferred solution to this would be Eos-side. Eos would maintain OSC in and out messages in a separate buffer from other log messages, so that when you turned off OSC logging, those messages would then disappear from the log, and when you turned it back on, they would reappear.
